File: arvores-python/tree.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def __str__(self):
        return str(self.data)

class BinaryTree:
    def __init__(self, data=None, node=None):
        if node:
            self.root = node
        elif data:
            self.root = Node(data)
        else:
            self.root = None
    
    def posOrderTraverse(self, node=None):
        if node is None:
            node = self.root
        if node.left:
            self.posOrderTraverse(node.left)
        if node.right:
            self.posOrderTraverse(node.right)
        print(node)

if __name__=="__main__":
    tree = BinaryTree()
    n1 = Node('3')
    n2 = Node('e')
    n3 = Node('5')
    n4 = Node('i')
    n5 = Node('r')
    n6 = Node('a')
    n7 = Node('n')
    n8 = Node('c')
    n9 = Node('v')
    n10 = Node('s')

    n1.left = n2
    n1.right = n3

    n2.left = n4
    n2.right = n5

    n3.left = n6

    n5.left = n7
    n5.right = n8

    n6.right = n9

    n8.right = n10

    tree.root = n1

    tree.posOrderTraverse()

File: trabalho/tree.c
#include <stdio.h>
#include <stdlib.h>

typedef struct n {
  int val;
  struct n* left;
  struct n* right;
} node;

typedef struct BinaryTree {
  node root;
} BinaryTree;

void preOrderPrint(node* root) {
  // node* pt = root;
  if (!root) return;
  printf("%d ", root->val);
  if (root->left != NULL) {
    preOrderPrint(root->left);
  }
  if (root->right != NULL) {
    preOrderPrint(root->right);
  }
}

void insere(node** root, int val) {
  if (*root == NULL) {
    node* new = (node*)malloc(sizeof(node));
    new->val = val;
    *root = new;
    new->left = NULL;
    new->right = NULL;
  }
  else {
    if (val > (*root)->val) {
      insere(&((*root)->right), val);
    }
    else if (val <= (*root)->val) {
      insere(&((*root)->left), val);
    }
  }
}

node* removeTree(node** root, int val) {
  node* x = *root;
  if (!x) {
    return x;
  }

  if (val == x->val) {
    // encontramos o no a ser removido
    if (!x->right && !x->left) {
      // no folha
      free(x);
      return NULL;
    }
    else if(!x->right || !x->left) {
      // no com 1 filho
      node* aux;
      if (x->right) 
        aux = x->right;
      else {
        aux = x->left;
      }
      free(x);
      return aux;
    }
    else {
      // no com dois filhos
      node* minNode = x->right;
      while(minNode->left) {
        minNode = minNode->left;
      }
      x->val = minNode->val;
      minNode->val = val;
      x->right = removeTree(&x->right, val);
    }
  }
  else if (val < x->val) {
    x->left = removeTree(&x->left, val);
  }
  else {
    x->right = removeTree(&x->right, val);
  }

  return x;

}

int main(int argc, char *argv[])
{
  node* root = NULL;
  char input[10];
  while(!feof(stdin)) {
    fgets(input, 10, stdin);
    switch (input[0]) {
      case 'i':
        insere(&root, atoi(&input[2]));
        break;
      case 'p':
        preOrderPrint(root);
        puts("");
        break;
      case 'r':
        removeTree(&root, atoi(&input[2]));
        break;
    }
  }

  return 0;
}
